packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -363,26 +404,56 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
   /// composing is not allowed.
   final int? maxLength;
 
+  /// Determines how the [maxLength] limit should be enforced.
+  ///
+  /// Defaults to [MaxLengthEnforcement.enforced].
+  ///
+  /// {@macro flutter.services.textFormatter.maxLengthEnforcement}
+  final MaxLengthEnforcement maxLengthEnforcement;
+
   /// Truncate the given TextEditingValue to maxLength user-perceived
   /// characters.
   ///
+  /// ### Composing text behaviors
+  ///
+  /// The composing range of [value] will be set to [TextRange.empty]
+  /// if [keepComposingRange] is false and the truncated value's length is more
+  /// than [value]'s. Otherwise only the part that exceeds [maxLength]
+  /// will be truncated.
+  ///
   /// See also:
   ///  * [Dart's characters package](https://pub.dev/packages/characters).
   ///  * [Dart's documentation on runes and grapheme clusters](https://dart.dev/guides/language/language-tour#runes-and-grapheme-clusters).
   @visibleForTesting
-  static TextEditingValue truncate(TextEditingValue value, int maxLength) {
+  static TextEditingValue truncate(
+    TextEditingValue value,
+    int maxLength, {
+    bool keepComposingRange = false,
+  }) {
     final CharacterRange iterator = CharacterRange(value.text);
     if (value.text.characters.length > maxLength) {
       iterator.expandNext(maxLength);
     }
     final String truncated = iterator.current;
+
+    final bool shouldKeepComposingRange = keepComposingRange &&
+      !value.composing.isCollapsed &&
+      truncated.length > value.composing.start &&
+      value.composing.start != math.min(value.composing.end, truncated.length);
```

Comment:
the last term seems redundant (i.e. can be inferred from   `!value.composing.isCollapsed && truncated.length > value.composing.start`)?

Reply:
Here might require more checking, I saw there is a getter `isNormalized` to check `end >= start`. Maybe we should check `truncated.length > math.max(value.composing.start, value.composing.end)` at here? Then the last condition can be inferred.

Reply:
`!value.composing.isCollapsed && truncated.length > math.max(value.composing.start, value.composing.end)`

Reply:
Hmm, an abnormal range won't coming in